bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/AmpPageImpl.java

```diff
@@ -0,0 +1,153 @@
+@Exporter(name = ExporterConstants.SLING_MODEL_EXPORTER_NAME,
+    extensions = ExporterConstants.SLING_MODEL_EXTENSION)
+public class AmpPageImpl implements AmpPage {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AmpPageImpl.class);
+
+    @SlingObject
+    private SlingHttpServletRequest request;
+
+    @ScriptVariable
+    private Page currentPage;
+
+    @OSGiService
+    private ClientLibraryAggregatorService aggregatorService;
+
+    private Map<String, String> pageLinkAttrs;
+    private List<String> headlibIncludes;
+
+    @PostConstruct
+    protected void init() {
+        String ampMode = AmpUtil.getAmpMode(request);
+        boolean isAmpSelector = Arrays.asList(request.getRequestPathInfo().getSelectors()).contains(AMP_SELECTOR);
+        pageLinkAttrs = new HashMap<>();
+        headlibIncludes = new ArrayList<>();
+        String relValue;
+        String hrefValue;
+        if (!isAmpSelector && ampMode.equals(AmpUtil.PAIRED_AMP)) {
+            relValue = "amphtml";
+            hrefValue = currentPage.getPath() + DOT + AMP_SELECTOR + URL_EXTENSION;
+        } else {
+            relValue = "canonical";
+            hrefValue = currentPage.getPath() + URL_EXTENSION;
+        }
+
+        pageLinkAttrs.put("rel", relValue);
+        pageLinkAttrs.put("href", hrefValue);
+        if (isAmpSelector) {
+            Set<String> resourceTypes = Utils.getResourceTypes(currentPage.getContentResource(),
+                aggregatorService.getResourceTypeRegex(), new HashSet<>());
+
+            try (ResourceResolver resolver = aggregatorService.getClientlibResourceResolver()) {
+                Utils.getTemplateResourceTypes(currentPage, aggregatorService.getResourceTypeRegex(), resolver,
+                    resourceTypes);
+                // Last part of any headlib path.
+                String headLibRelPath = "/" + aggregatorService.getHeadlibName();
+
+                // Iterate through each resource type and read its AMP headlib.
+                for (String resourceType : resourceTypes) {
+
+                    // Resolve the resource type's AMP headlib.
+                    Resource headLibResource;
+                    String headLibPath = resourceType + headLibRelPath;
+                    headLibResource = Utils.resolveResource(resolver, headLibPath);
+                    if (headLibResource == null) {
+                        LOG.trace("No custom headlib for resource type {}.", resourceType);
+
+                        // Get headLibResource from resource superType.
+                        headLibResource = getHeadlibResourceSuperType(resolver, resourceType, headLibRelPath);
+                        if (headLibResource == null) {
+                            LOG.trace("No custom headlib for resource superType from resource type {}.", resourceType);
+                            continue;
+                        }
+                    }
+                    headlibIncludes.add(headLibResource.getPath());
```

Comment:
@id-keenan , I believe we still need to dedupe the list of paths. The current JS aggregation of headlibs does not seem to dedupe the paths here so this would end up spitting out several time the JS library if there are multiple times the same resource type on the page.